cpp/src/replace/replace.cu

```diff
@@ -77,47 +102,97 @@ namespace{ //anonymous
+                                             static_cast<col_type*>(output.data),
+                                             typed_out_valid,
+                                             null_count,
                                              old_values_begin,
-                                             old_values_begin + nvalues,
-                                             static_cast<const col_type*>(d_new_values));
+                                             old_values_begin + new_values.size,
+                                             static_cast<const col_type*>(new_values.data),
+                                             typed_new_valid);
+      if (input_is_nullable || new_is_nullable) {
+        CUDA_TRY(cudaMemcpy(&output.null_count, null_count,
+                               sizeof(gdf_size_type), cudaMemcpyDefault));
+        RMM_FREE(null_count, 0);
+      }
     }
   };
 
-  gdf_error find_and_replace_all(gdf_column*       col,
-                                 const gdf_column* old_values,
-                                 const gdf_column* new_values)
+  gdf_column find_and_replace_all(const gdf_column  &input_col,
+                                  const gdf_column &old_values,
+                                  const gdf_column &new_values)
   {
-    GDF_REQUIRE(col != nullptr && old_values != nullptr && new_values != nullptr, GDF_DATASET_EMPTY);
-    GDF_REQUIRE(old_values->size == new_values->size, GDF_COLUMN_SIZE_MISMATCH);
-    GDF_REQUIRE(col->dtype == old_values->dtype && col->dtype == new_values->dtype, GDF_DTYPE_MISMATCH);
-    GDF_REQUIRE(col->valid == nullptr || col->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
-    GDF_REQUIRE(old_values->valid == nullptr || old_values->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
-    GDF_REQUIRE(new_values->valid == nullptr || new_values->null_count == 0, GDF_VALIDITY_UNSUPPORTED);
-
-    
-    cudf::type_dispatcher(col->dtype, replace_kernel_forwarder{},
-                          col->data,
-                          col->size,
-                          old_values->data,
-                          new_values->data,
-                          new_values->size); 
-
-    return GDF_SUCCESS;
+    CUDF_EXPECTS(input_col.data != nullptr, "Null input data.");
+    CUDF_EXPECTS(old_values.data != nullptr && new_values.data != nullptr, "Null replace data.");
+    CUDF_EXPECTS(old_values.size == new_values.size, "old_values and new_values size mismatch.");
+    CUDF_EXPECTS(input_col.dtype == old_values.dtype && input_col.dtype == new_values.dtype, "Columns type mismatch.");
+    CUDF_EXPECTS(old_values.valid == nullptr || old_values.null_count == 0, "Nulls can not be replaced.");
+
+    gdf_column output = cudf::empty_like(input_col);
+
+    gdf_size_type column_byte_width{gdf_dtype_size(input_col.dtype)};
+
+    void *data = nullptr;
+    gdf_valid_type *valid = nullptr;
+    RMM_ALLOC(&data, input_col.size * column_byte_width, 0);
```

Comment:
Why not use `cudf::allocate_like` instead?

Reply:
Can we use `cudf::copy()` ?  then replace the required values.

Reply:
My intuition is that that would be more expensive since the `copy` invokes a kernel to copy the data over. Then another kernel to do the replacement. Currently, you're doing the copy and the replacement in a single kernel which fuses the two operations.

Reply:
But currently, it uses too many `if-then-else` statements which may cause branch divergence.

Reply:
Branch divergence is unavoidable in replacing the null values. However, you can nicely avoid the branch divergence when the columns do not contain null values via the `bool` template args.